pkg/hub/taints/controller.go

```diff
@@ -0,0 +1,166 @@
+	recorder events.Recorder) factory.Controller {
+	c := &taintsController{
+		kubeClient:    kubeClient,
+		clusterClient: clusterClient,
+		clusterLister: clusterInformer.Lister(),
+		eventRecorder: recorder.WithComponentSuffix("taints-controller"),
+	}
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, clusterInformer.Informer()).
+		WithSync(c.sync).
+		ToController("taintsController", recorder)
+}
+
+func (c *taintsController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	managedClusterName := syncCtx.QueueKey()
+	klog.V(4).Infof("Reconciling Taints %s", managedClusterName)
+	managedCluster, err := c.clusterLister.Get(managedClusterName)
+	if errors.IsNotFound(err) {
+		// Spoke cluster not found, could have been deleted, do nothing.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	managedCluster = managedCluster.DeepCopy()
+	newTaints := managedCluster.Spec.Taints
+
+	switch {
+	case meta.IsStatusConditionTrue(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable):
+		newTaints, _ = c.deleteTaintAndJudgeExist(newTaints, "",
+			v1.ManagedClusterTaintUnreachable, v1.ManagedClusterTaintUnavailable)
+	case meta.IsStatusConditionFalse(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable):
+		var exist bool
+		newTaints, exist = c.deleteTaintAndJudgeExist(newTaints,
+			v1.ManagedClusterTaintUnavailable, v1.ManagedClusterTaintUnreachable)
+		if !exist {
+			t1 := v1.Taint{
+				Key:    v1.ManagedClusterTaintUnavailable,
+				Value:  ConditionFalse,
+				Effect: v1.TaintEffectNoSelect,
+				TimeAdded: metav1.Time{
+					Time: time.Now(),
+				},
+			}
+			newTaints = append(newTaints, t1)
+		}
+	case c.isIsStatusConditionUnknown(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable) ||
+		meta.FindStatusCondition(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable) == nil:
+		var exist bool
+		newTaints, exist = c.deleteTaintAndJudgeExist(newTaints,
+			v1.ManagedClusterTaintUnreachable, v1.ManagedClusterTaintUnavailable)
+		value := ConditionUnknown
+		if meta.FindStatusCondition(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable) == nil {
+			value = NoCondition
+		}
+		if !exist {
+			t1 := v1.Taint{
+				Key:    v1.ManagedClusterTaintUnreachable,
+				Value:  value,
+				Effect: v1.TaintEffectNoSelect,
+				TimeAdded: metav1.Time{
+					Time: time.Now(),
+				},
+			}
+			newTaints = append(newTaints, t1)
+		}
+	}
+	if c.isTaintsEqual(managedCluster.Spec.Taints, newTaints) {
+		return nil
+	}
+
+	managedCluster.Spec.Taints = newTaints
+	managedCluster, err = c.clusterClient.ClusterV1().ManagedClusters().Update(ctx, managedCluster, metav1.UpdateOptions{})
+	return err
+}
+
+func (c *taintsController) deleteTaintAndJudgeExist(taints []v1.Taint, isExistKey string, deleteKeys ...string) ([]v1.Taint, bool) {
```

Comment:
would you split this method into two methods? one is used to delete and the other is used to judge exist, I think it will be more clear